[SPARK-23890][SQL] Support DDL for adding nested columns to struct types

ottomata opened the pull request:
## What changes were proposed in this pull request?

SPARK-23525 added the ability to `ALTER TABLE CHANGE COLUMN` for simple COMMENT only changes.  For safety, column changes are generally restricted.  However, `ALTER TABLE ADD COLUMN` is safe and allowed.  In order to add columns to a nested struct type, we must run an `ALTER TABLE CHANGE COLUMN` command, since struct type DDL look like single top level columns with a complex type.

Given on origin column declared like
```
nested struct<s1:string>
```

This patch allows users to issue SQL DDL like:

```
ALTER TABLE t1 CHANGE COLUMN nested nested struct<s1:string,s2:string>
```

to add sub columns to a struct.  It does this by recursing over StructTypes, and comparing dataTypes for each shared field between the origin table and the new destination type.

## How was this patch tested?

`testChangeColumn` in DDLSuite.scala and was amended and tests were added
to HiveDDLSuite.scala  to include tests for altering struct type columns.

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

gatorsmile commented:
@ottomata Thanks for contributing this! Could you verify the behaviors of our built-in data sources (e.g. Parquet, ORC) and see whether they work properly after adding a field into a nested struct?

ottomata commented:
Ya will do!  Actually this patch (from long ago) has a bug and fails a test.  I have another one half finished to fix that, will try to get back to it soon. 

> On Dec 22, 2018, at 20:09, Xiao Li <notifications@github.com> wrote:
> 
> @ottomata Thanks for contributing this! Could you verify the behaviors of our built-in data sources (e.g. Parquet, ORC) and see whether they work properly after adding a field into a nested struct?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.

ottomata committed:
[SPARK-23890][SQL] Support ALTER TABLE CHANGE COLUMN to add fields to structs

SPARK-23525 added the ability to ALTER TABLE CHANGE COLUMN for
simple COMMENT only changes.  For safety, column changes are
generally restricted.  However, ALTER TABLE ADD COLUMN is safe
and allowed.  In order to add columns to a nested struct type,
we must run an ALTER TABLE CHANGE COLUMN command, since
struct type DDL look like single top level columns with
a complex type.

Given on origin column declared like
  nested struct<s1:string>

This patch allows users to issue SQL DDL like:

  ALTER TABLE t1 CHANGE COLUMN nested nested struct<s1:string,s2:string>

to add sub columns to a struct.  It does this by recursing over
StructTypes, and comparing dataTypes for each shared field between
the origin table and the new destination type.

testChangeColumn in DDLSuite.scala and was amended and tests were added
to HiveDDLSuite.scala  to include tests for altering struct type columns.

ottomata commented:
@gatorsmile, I've updated with my fix (the previous code lost column metadata), and rebased on top of master.

I know adding columns top level and in structs works for Parquet (we do this all the time at WMF).  Is there a more robust unit test in Spark somewhere that I can verify this with?  We don't use ORC, but I'll see if I can experiment with it to verify it works there.

jobar commented on sql/core/src/main/scala/org/apache/spark/sql/execution/command/ddl.scala:
```diff
@@ -294,8 +295,10 @@ case class AlterTableUnsetPropertiesCommand(
 /**
- * A command to change the column for a table, only support changing the comment of a non-partition
+ * A command to change the column for a table.  For now, this only supports:
  * column for now.
```
This line should be removed.

jobar commented on sql/core/src/test/scala/org/apache/spark/sql/execution/command/DDLSuite.scala:
```diff
@@ -1693,16 +1698,41 @@ abstract class DDLSuite extends QueryTest with SQLTestUtils {
+      sql("ALTER TABLE dbx.tab1 CHANGE COLUMN col2 col2 INT")
+    }
+
+    // Ensure that change column with in nested column with type change fails
```
Typo plus not clear-  I'd say: `Ensure that change column changing type within nested-column fails`

jobar reviewed (commented):
Minor comments in comments  - Looks good :)

ottomata committed:
Comment fixes

ottomata commented:
@gatorsmile, I just verified that this works for ORC too:

```sql
> CREATE TABLE orc0 (
  name string,
  val int,
  s struct<v1:string>
) STORED AS ORC;

> insert into table orc0 select "a", 1, named_struct("v1", "V1");

> select * from orc0;
name    val s
a   1   {"v1":"V1"}

> alter table orc0 change column s s struct<v1:string,v2:string>;

> select * from orc0;
name    val s
a   1   {"v1":"V1","v2":null}

> insert into table orc0 select "a", 1, named_struct("v1", "V1", "v2", "V2");

> select * from orc0;
name    val s
a   1   {"v1":"V1","v2":null}
a   1   {"v1":"V1","v2":"V2"}
```

This was using Hive 1.1.0.

ottomata commented:
Looking for a reviewer here!  Ping @srowen, can you review, or recommend someone else to ask?

Thanks!

srowen commented:
This one isn't my area; @gatorsmile, @cloud-fan perhaps

ottomata commented:
@dongjoon-hyun, do you have any suggestions on how I can motivate someone to review this? :)

yyhpys commented:
@ottomata Hi. I have been using your PR in a production thrift server. Really appreciate for the contribution.
But since there was an type comparison issue with ArrayType field, I am using it with the following additional modification.

```scala
  private def columnsCompatible(
    ours: StructField,
    theirs: StructField,
    resolver: Resolver
  ): Boolean = {
    // If we can't resolve between the column names, then these are not compatible.
    if (!resolver(ours.name, theirs.name)) {
      return false
    }
    //match dataType
    typeCompatible(ours.dataType, theirs.dataType, resolver)
  }
  
  private def typeCompatible(
    ourType: DataType,
    theirType: DataType,
    resolver: Resolver
  ): Boolean = {
    (ourType, theirType) match {
      // If dataTypes are the same, then these columns are compatible
      case (a, b) if a == b => true
      // If both are StructTypes, recurse and ensure that each of
      // the common fields recursively have identical types.
      case (ourSchema: StructType, theirSchema: StructType) =>
        // Get all fields in theirSchema that are also in ourSchema
        // and make sure that all of these fields are also compatible.
        theirSchema
        .filter(field => findColumnByName(ourSchema, field.name, resolver).isDefined)
        .forall { theirField =>
          columnsCompatible(
            findColumnByName(ourSchema, theirField.name, resolver).get,
            theirField,
            resolver
          )
        }
      case (ourSchema: ArrayType, theirSchema: ArrayType) =>
        // if both are ArrayTypes, then match their element type
        typeCompatible(ourSchema.elementType, theirSchema.elementType, resolver)
      case _ => false
    }
  }
```

Hope you consider this case and reinforce this PR if possible. Thanks again.

ottomata committed:
[SPARK-23890][SQL] Support ALTER TABLE CHANGE COLUMN to add fields struct elements in arrays

Thanks for the comment @yyhpys:
https://github.com/apache/spark/pull/21012#issuecomment-511692305

ottomata committed:
Merge branch 'master' into SPARK-23890

cloud-fan commented:
AFAIK we already support changing nested fields with command `AlterTable`. `AlterTableChangeColumnCommand` is outdated and will be removed in the future. cc @rdblue

ottomata commented:
@cloud-fan oh cool!  Can you tell me how to do this then?  I just tried with Spark 2.4.3:

```sql
CREATE TABLE otto.p1 (
    name string,
    val int,
    s struct<v1:string>
) STORED AS PARQUET;
```
```scala
scala> spark.sql("alter table otto.p1 change column s s struct<v1:string,v2:string>")
org.apache.spark.sql.AnalysisException: ALTER TABLE CHANGE COLUMN is not supported for changing column 's' with type 'StructType(StructField(v1,StringType,true))' to 's' with type 'StructType(StructField(v1,StringType,true), StructField(v2,StringType,true))';
  at org.apache.spark.sql.execution.command.AlterTableChangeColumnCommand.run(ddl.scala:323)
...
```

cloud-fan commented:
it's only in master, and is only applicable to the data source v2 catalog. You can try
```
SET spark.sql.catalog.cat1 = com.my.TheImplClass
CREATE TABLE cat1.t ...
ALTER TABLE cat1.t CHANGE COLUMN ...
```
`TheImplClass` should implement `TableCatalog`. We will migrate the builtin hive catalog to use the new API in the future, so that we can remove `AlterTableChangeColumnCommand`.

See the tests in `DataSourceV2SQLSuite`

ottomata commented:
Ah nice, that looks great.  I gather these are part of [SPARK-22386](https://jira.apache.org/jira/browse/SPARK-22386), which targets Spark 3.0.0.  Guess I gotta wait a while longer then, eh? :)

rdblue commented:
Yes, I think we should remove the AlterTableChangeColumnCommand command in the future. I'm not sure when that will be, though. We don't want to make that breaking change for 3.0, so it will likely persist until 4.0

ottomata commented:
I don't mind either way, as long as I can alter tables via spark.sql.
Right now I'm using the Hive API and opening up a direct connection to Hive
to bypass this problem. Wikimedia only uses stable releases though, so I
have to wait until the v2 catalog improvements land.

On Thu, Aug 29, 2019 at 1:11 PM Ryan Blue <notifications@github.com> wrote:

> Yes, I think we should remove the AlterTableChangeColumnCommand command in
> the future. I'm not sure when that will be, though. We don't want to make
> that breaking change for 3.0, so it will likely persist until 4.0
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/spark/pull/21012?email_source=notifications&email_token=AAMM7RQUCPMHEKVUAVAGCZTQG77MDA5CNFSM4EZUJFZ2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD5PFVXA#issuecomment-526277340>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAMM7RW7YJ3DBU7UFPLWIJDQG77MDANCNFSM4EZUJFZQ>
> .
>

AmplabJenkins commented:
Can one of the admins verify this patch?

ottomata commented:
Hi, I'd like to reopen this issue.  Back in 2019, @cloud-fan said that v2 datasource:
> already support changing nested fields with command AlterTable 

However, there is an [explicit check](https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala#L1441) and [test](https://github.com/apache/spark/blob/e3f46ed57dc063566cdb9425b4d5e02c65332df1/sql/core/src/test/scala/org/apache/spark/sql/connector/AlterTableTests.scala#L583) that prevents this from happening.

We need to be able to evolve tables in a backwards compatible way.  This safe and possible with top level fields, and it should be safe and possible with struct column fields too, as long as we are only adding new fields.  

cc @MaxGekk @beliefer @brkyvz 

Thanks!

ottomata commented:
https://issues.apache.org/jira/browse/SPARK-23890

cloud-fan commented:
With DS v2, I think you can use `ADD COLUMN` to add nested fields?

ottomata commented:
@cloud-fan oh!  THIS WORKS!

```sql
spark-sql (default)> describe otto.test_table03;
col_name	data_type	comment
s1	struct<s1_f1:string>

spark-sql (default)> ALTER TABLE otto.test_table03 ADD COLUMN s1.s1_f2_added STRING;

'spark-sql (default)> describe otto.test_table03;
col_name	data_type	comment
s1	struct<s1_f1:string,s1_f2_added:string>
```

AMAZING!  Thank you! Feels simpler than `change column`.  <3

ottomata commented:
@cloud-fan in my tests, this works for tables created with e.g. `USING ICEBERG`, but not with tables `USING PARQUET` or `USING HIVE`.  Altering a nested column with HIVE and PARQUET both throw e.g.

```
Error in query: AlterTableAddColumnsCommand does not support nested column: s1.added_field0
```

We have a lot of existent Hive parquet tables we'd like evolve.  We would be fine with losing support for querying them with Hive MR (relying only on spark and presto), but the Hive metastore is pretty crucial.  Will Spark support altering nested columns on Hive parquet tables?

I'm sure there is something I'm missing here with Spark catalogs, etc.  FWIW, we are on Spark 3.1.2 right now.

ottomata commented:
[Verified that the behavior is the same in Spark 3.5](https://phabricator.wikimedia.org/T209453#9413866).

cloud-fan commented:
To support this in DS v1, we need to understand the consequence. Parquet is probably fine adding a nested field, but how about other data sources like JSON, CSV, etc.?

ottomata commented:
>  how about other data sources like JSON, CSV, etc.?

From a theoretical perspective JSON would be fine.  
Are nested fields even supported in CSV?

[Here](https://github.com/apache/spark/pull/21012#issuecomment-474054663) I tested that ORC supports this.

> To support this in DS v1

I've tried googling and reading code, but was not able to totally understand how this works.  How does Spark know that e.g. a Hive table should be DS v1 or DSv2?  

I had thought that if i created the table with `USING` [the Spark Data Source syntax](https://spark.apache.org/docs/3.0.0/sql-ref-syntax-ddl-create-table-datasource.html), it would be DSv2, but in my [tests it was not](https://phabricator.wikimedia.org/T209453#9413866).

cloud-fan commented:
We should revisit the code here and determine which file sources support adding nested fields:
https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/command/tables.scala#L266-L278

ottomata commented:
> We should revisit the code here and determine which file sources support adding nested fields
I'd like to help, but I'm not sure how.  Do you mean theoretically which file sources can support adding nested fields?  I believe all of those should, except perhaps CSV (since I don't think CSV supports nested fields?).  

Or, do you mean examine the spark DataSource code and make sure the underlying data files can be read with an altered schema that has extra nested columns? If so, I have confirmed that this does work with JSON, ORC and Parquet datasource V1.

cloud-fan commented:
We can update `AlterTableAddColumnsCommand` to support nested columns, with a allowlist for file sources that definitely support it (parquet, orc, json)

ottomata commented:
Okay, I think I see.  So, tests and a patch to do this.

I've reopened https://issues.apache.org/jira/browse/SPARK-23890 to accommodate.

ottomata commented:
Starting to try this.  I think there is more to change than just the list of datasources for which `verifyAlterTableAddColumn` will throw `alterAddColNotSupportDatasourceTableError`.

In this [commit](https://github.com/apache/spark/compare/master...ottomata:spark:SPARK-23890-add-nested-column), I've done so, but now a check is failing in ResolveCatalogSession.  It looks like even with `USING json` or `USING hive`, [this case match](https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveSessionCatalog.scala#L55-L60) against `ResolvedV1TableIdentifier` is matching, causing 

```
[UNSUPPORTED_FEATURE.TABLE_OPERATION] The feature is not supported: Table `spark_catalog`.`default`.`t1` does not support ADD COLUMN with qualified column. Please check the current catalog and namespace to make sure the qualified table name is expected, and also check the catalog implementation which is configured by "spark.sql.catalog". SQLSTATE: 0A000
```

to be thrown.

@cloud-fan, 
Do we need to an another case in ResolveCatalogSession before this one to also allow these v2 datasources, even though they have v1 compatibility TableIdentifiers (?) to continue with the ALTER command?

ottomata commented:
Ah, sorry, my scala goggles weren't on quite right.  According to `verifyAlterTableAddColumn`, e.g. JsonDataSourceV2 is supported for ALTER TABLE ADD COLUMN.  We don't need to change that check at all.

What we need to change is the check in ResolveCatalogSession that does
```scala
throw QueryCompilationErrors.unsupportedTableOperationError(ident, "ADD COLUMN with qualified column")
```

`if c.name.length > 1`.  
However, if I remove that, I can see a problem in the way `AlterTableAddColumnsCommand` is implemented: it takes a `Seq[StructField]` as the list of columns to add.  This list is constructed by `convertToStructField` in ResolveSessionCatalog, which naively assumes that the col to add is not nested: 
```scala
    StructField(col.name.head, col.dataType, nullable = true, builder.build())
```
Returns a StructField with the top level column name but the nested leaf column's DataType.

ottomata commented:
Okay, @cloud-fan I got something working, but I'm not sure if you are going to like it!  I'd like to vet it with you before I keep working on it.

`convertToStructField` now [returns a nested StructField](https://github.com/apache/spark/commit/d00ac40a4699b04632367d863b5e4bdbee481348#diff-415c3b78d7558933c759da01218075e011c45c4d0fb5ff3fe522f24d6e4b8e20R608) if `col.name.length > 1`. Then, in `AlterTableAddColumnsCommand`, The `colsToAdd`, which might be nested structs now, is [merged as a StructType schema with the existing table schema](https://github.com/apache/spark/commit/d00ac40a4699b04632367d863b5e4bdbee481348#diff-632efff0ffceb2ff47b97f39570b4f2cff24401f132f559158acaa651b27af26R244).  This gets us a full new schema with new nested columns recursively merged in.  Duplicate column names are still checked for, but now only leaf node path names are checked.  This works because the merge merges together struct columns from the colsToAdd with existing struct columns, so we know there are no duplicate container/struct columns in the new schema.

[SchemaUtils.merge](https://github.com/apache/spark/commit/d00ac40a4699b04632367d863b5e4bdbee481348#diff-79df452a8a9b284ece3d9b034058e5fcbb5a421890e4947926963e624b7be1d5R331) needs the most work.  It's based on Wikimedia code we currently use to merge schemas together, but as is it will only work in case-insensitive catalogs (Hive).  There are probably other things to deal with in review.

If you are okay with this direction, I'll reopen this PR with this code.

---
Question though.  This original PR for DataSource v1 via `ALTER TABLE CHANGE COLUMN` worked with adding columns to structs in ArrayType element type, and should also work for structs in MapType value types too.  I can address a specific qualified column (e.g. `a.b.c`) in a struct with `ADD COLUMNS`, but I don't think we address the element type of an array column or the value type of a map column in this way.  Would one need to use `CHANGE|ALTER COLUMN` syntax for this?  And if so, should that be supported?...And if so, would it not be easiest to add columns struct fields via `CHANGE|ALTER COLUMN` too?

ottomata commented:
Ping @cloud-fan

ottomata commented:
> Would one need to use CHANGE|ALTER COLUMN syntax for this?

[TIL](https://phabricator.wikimedia.org/T209453#10632894) that Iceberg supports this with .value column name referencing!

https://iceberg.apache.org/docs/nightly/spark-ddl/#alter-table-add-column

```sql
-- create a map column of struct key and struct value
ALTER TABLE prod.db.sample
ADD COLUMN points map<struct<x: int>, struct<a: int>>;

-- add a field to the value struct in a map. Using keyword 'value' to access the map's value column.
ALTER TABLE prod.db.sample
ADD COLUMN points.value.b int;
```

And, that works via Spark!  

I guess there is a bug though:
https://github.com/apache/iceberg/issues/2962
